src/mscorlib/src/System/String.cs

```diff
@@ -568,12 +568,14 @@ private unsafe static int CompareOrdinalHelper(String strA, String strB)
 #endif
                 
                 DiffNextInt:
-                int order;
-                if ( (order = (int)*a - (int)*b) != 0) {
-                    return order;
-                }
-                Contract.Assert( *(a+1) != *(b+1), "This char must be different if we reach here!");
-                return ((int)*(a+1) - (int)*(b+1));
+                if (*a == *b) goto ReturnDifference;
+
+                DiffOffset1:
+                Contract.Assert(*(a + 1) != *(b + 1), "This char must be different if we reach here!");
+                a += 1; b += 1;
```

Comment:
Have you tried to measuring the different options here? 

I would expect that this is going to have extra memory fetch that should cost a few extra cycles (many iterations are needed to measure a few cycle differences).

I would expect that the following is going to work better:

```
DiffNextInt:
   if (*a == *b)
        return *a - *b;

DiffOffset1:
  Contract.Assert(...);
  return *(a + 1) - *(b + 1);
```